src/coins.cpp

```diff
@@ -914,7 +916,7 @@ CAmount CCoinsViewCache::GetValueIn(const CTransaction& tx) const
     return nResult;
 }
 
-bool CCoinsViewCache::HaveShieldedRequirements(const CTransaction& tx) const
+std::optional<UnsatisfiedShieldedReq> CCoinsViewCache::HaveShieldedRequirements(const CTransaction& tx) const
```

Comment:
Is this how we want to represent the equivalent of a `Result` type? It seems potentially error-prone that it boolean-coerces to `false` on success and `true` on failure.

(There was a [proposal to add std::expected](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4015.pdf) but it never went anywhere, so I think we'd want to add our own, either based on that or based on Rust's [std::result](https://doc.rust-lang.org/std/result/). There is also a [similar type in Boost](https://www.boost.org/doc/libs/1_70_0/libs/outcome/doc/html/tutorial/essential/result.html) but I guess we don't want to use that.)

Reply:
I would definitely prefer something more similar to a Rust `Result` for expected errors, and we could add that in a subsequent refactor.

Reply:
Filed #4816 .